binding.cc

```diff
@@ -836,13 +891,32 @@ struct CloseWorker final : public BaseWorker {
   CloseWorker (napi_env env,
                Database* database,
                napi_value callback)
-    : BaseWorker(env, database, callback, "leveldown.db.close") {}
+    : BaseWorker(env, database, callback, "leveldown.db.close") {
+      // Visit iterators on the main thread and prevent GC so that we can safely
+      // visit them again in the asynchronous worker to close them explicitly.
+      // Mark as closing to disallow further operations to be scheduled.
+      database_->VisitIterators([&] (Iterator* iterator) {
+        iterator->IncrementReferenceCount();
+        iterator->closing_ = true;
+      });
+    }
 
   ~CloseWorker () {}
 
   void DoExecute () override {
+    database_->VisitIterators([&] (Iterator* iterator) {
+      iterator->Close();
```

Comment:
Hmm, GC-ed iterators are closed on the main thread so might as well do the same\* here. For this PR to be acceptable, we need closing to be fast anyway (AFAIK there's no disk access involved in e.g. releasing the snapshot so it might be fine).

\* I.e. move the `iterator->Close()` call up, which means I can then also replace usage of `closing_` with `closed_`, move the `Detach()` call back into `Close()`, and forego the reference counting (possibly even altogether, because it is not strictly necessary while nexting, which is already guarded from GC by the fact that `next()` callback in JS land has a reference to the iterator). Jeez :)

Reply:
Reconsidered. On database close, there may be multiple iterators left (with references in userland) so I prefer to do the work in a worker pool thread. And I was wrong about the iterator always being persistent while nexting (there's a window in between `NextWorker::DoComplete` and `::DoFinally` where it can be GC-ed, I added a test for this edge case).